enhancements/installer/vsphere-ipi-multiple-datacenter-clusters.md

```diff
@@ -0,0 +1,1169 @@
+
+  // Datacenter in which VMs are located.
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MinItems=1
+  Datacenters []string `json:"datacenters"`
+
+  // Soap round tripper count (retries = RoundTripper - 1)
+  // +kubebuilder:validation:Optional
+  RoundTripperCount uint `json:"soapRoundtripCount,omitempty"`
+
+  // Specifies the path to a CA certificate in PEM format. Optional; if not
+  // configured, the system's CA certificates will be used.
+  // +kubebuilder:validation:Optional
+  CAFile string `json:"caFile,omitempty"`
+
+  // Thumbprint of the VCenter's certificate thumbprint
+  // +kubebuilder:validation:Optional
+  Thumbprint string `json:"thumbprint,omitempty"`
+
+  // Name of the secret where vCenter credentials are present.
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  // +kubebuilder:validation:Required
+  SecretName string `json:"secretName"`
+
+  // Namespace where the secret will be present containing vCenter credentials.
+  // +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=63
+  // +kubebuilder:validation:Required
+  SecretNamespace string `json:"secretNamespace"`
+
+  // IP Family enables the ability to support IPv4 or IPv6
+  // Supported values are:
+  // ipv4 - IPv4 addresses only (Default)
+  // ipv6 - IPv6 addresses only
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:MaxItems=2
+  IPFamilyPriority []string `json:"ipFamily,omitempty"`
+}
+
+// VSpherePlatformPlacementConstraint is the context information for VM placements within a failure domain
+type VSpherePlatformPlacementConstraint struct {
+  // resourcePool is the absolute path of the resource pool where virtual machines will be
+  // created. The absolute path is of the form /<datacenter>/host/<cluster>/Resources/<resourcepool>.
+  // TODO: maximum length?
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:MaxLength=1024
+  ResourcePool string `json:"resourcePool,omitempty"`
+
+  // folder is the name or inventory path of the folder in which the
+  // virtual machine is created/located.
+  // TODO: maximum length?
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:MaxLength=1024
+  Folder string `json:"folder,omitempty"`
+}
+
+// VSpherePlatformDeploymentZoneSpec holds the association between a
+// vCenter, failure domain and the virtual machine placementConstraints
+type VSpherePlatformDeploymentZoneSpec struct {
+
+  // TODO: name is the name of the VSpherePlatformDeploymentZoneSpec
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  // +kubebuilder:validation:Required
+  Name string `json:"name"`
+
+  // server is the fully-qualified domain name or the IP address of the vCenter server.
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MaxLength=255
```

Comment:
Where does this max length come from, if this is a FQDN, surely it can be quite long with a bunch of different subdomains? Or is this a limit of the DNS spec?